### ISSUE TITLE
Learning rate adaptation

### ISSUE THREAD

**Issue by Some-random** [espnet/espnet#1583]
I've tried the learning rate adaptation schedule in this script (https://github.com/espnet/espnet/blob/master/egs/hkust/asr1/run.sh) and found it confounding... 

My two questions are:
1 Why are we not decreasing learning rate (like divide warmup_k by 10) throughout training?
2 What's the intuition behind average_checkpoint operation in decoding (like a paper?) and why is it getting much better results compared to single best checkpoint on dev set?

**Comment by sw005320** [comment]
> I've tried the learning rate adaptation schedule in this script (https://github.com/espnet/espnet/blob/master/egs/hkust/asr1/run.sh) and found it confounding...
> 
> My two questions are:
> 1 Why are we not decreasing learning rate (like divide warmup_k by 10) throughout training?

We have learning rate decay, e.g., for transformer https://github.com/espnet/espnet/blob/10d36929184606137c8a7b4b9ad9e3cc624e0c6a/espnet/scheduler/scheduler.py#L116
We also have learning rate decay in RNN through the epsilon decay.

> 2 What's the intuition behind average_checkpoint operation in decoding (like a paper?) and why is it getting much better results compared to single best checkpoint on dev set?

It is not theoretically grounded for non-linear classifiers, but in general averaging operations makes the model more generalize (especially when the model is well converged). You could find a lot of reports in NMT. Kaldi also uses it.

**Comment by Some-random** [comment]
> > I've tried the learning rate adaptation schedule in this script (https://github.com/espnet/espnet/blob/master/egs/hkust/asr1/run.sh) and found it confounding...
> > My two questions are:
> > 1 Why are we not decreasing learning rate (like divide warmup_k by 10) throughout training?
> 
> We have learning rate decay, e.g., for transformer
> 
> https://github.com/espnet/espnet/blob/10d36929184606137c8a7b4b9ad9e3cc624e0c6a/espnet/scheduler/scheduler.py#L116
> 
> 
> We also have learning rate decay in RNN through the epsilon decay.
> > 2 What's the intuition behind average_checkpoint operation in decoding (like a paper?) and why is it getting much better results compared to single best checkpoint on dev set?
> 
> It is not theoretically grounded for non-linear classifiers, but in general averaging operations makes the model more generalize (especially when the model is well converged). You could find a lot of reports in NMT. Kaldi also uses it.

Thanks for the answer to question 2! Perhaps I wasn't clear enough on question 1. The two learning rate schedules I was comparing were **divide warmup_k by 10 whenever loss doesn't decrease on dev set** v.s. **stick to original warmup_k and just wait for learning rate to decrease slowly**.

And I actually tried using the latter on HKUST and decoding with average_checkpoint, but got higher WER than the former... It seems counter-intuitive to me because the former will yield a smaller learning rate at the end of training and judging by performance on dev set, the former does get a lower WER...

**Comment by lyqlola** [comment]
@Some-random Check this paper [Exploring RNN-Transducer for Chinese speech recognition](https://arxiv.org/abs/1811.05097). The model may overfit the data during training and a more aggressive learning decay strategy, ie. dividing the lr by 10, will help the model to overcome the overfitting problem.
